File: main/client/test/src/mill/main/client/RetryRule.java
// Taken from https://www.swtestacademy.com/rerun-failed-test-junit/
package mill.main.client;
import java.util.Objects;
import org.junit.rules.TestRule;
import org.junit.runner.Description;
import org.junit.runners.model.Statement;
public class RetryRule implements TestRule {
    private int retryCount;
    public RetryRule(int retryCount) {
        this.retryCount = retryCount;
    }
    public Statement apply(Statement base, Description description) {
        return statement(base, description);
    }
    private Statement statement(final Statement base, final Description description) {
        return new Statement() {
            @Override
            public void evaluate() throws Throwable {
                Throwable caughtThrowable = null;
                // implement retry logic here
                for (int i = 0; i < retryCount; i++) {
                    try {
                        base.evaluate();
                        return;
                    }
                    catch (Throwable t) {
                        caughtThrowable = t;
                        System.err.println(description.getDisplayName() + ": run " + (i + 1) + " failed.");
                    }
                }
                System.err.println(description.getDisplayName() + ": Giving up after " + retryCount + " failures.");
                throw Objects.requireNonNull(caughtThrowable);
            }
        };
    }
}

File: main/client/test/src/mill/main/client/RetryRule.scala
package mill.main.client

import org.junit.rules.TestRule
import org.junit.runner.Description
import org.junit.runners.model.Statement
import java.util.Objects

class RetryRule(retryCount: Int) extends TestRule {

  override def apply(base: Statement, description: Description): Statement = {
    new Statement {
      override def evaluate(): Unit = {
        var caughtThrowable: Throwable = null
        // implement retry logic here
        for (i <- 0 until retryCount) {
          try {
            base.evaluate()
            return
          } catch {
            case t: Throwable =>
              caughtThrowable = t
              System.err.println(s"${description.getDisplayName}: run ${i + 1} failed.")
          }
        }
        System.err.println(s"${description.getDisplayName}: Giving up after $retryCount failures.")
        throw Objects.requireNonNull(caughtThrowable)
      }
    }
  }
}
